transport-native-epoll/src/main/java/io/netty/channel/epoll/TunTapPacketLogger.java

```diff
@@ -0,0 +1,217 @@
+        }
+
+        // Print the full contents of the packet.
+        if (logPacketData) {
+            formatter.format("%s", ByteBufUtil.prettyHexDump(packetData));
+        }
+
+        formatter.close();
+
+        return out.toString();
+    }
+
+    private static void formatEthernetHeader(ByteBuf packetData, int headerOffset, int protocol, Formatter formatter) {
+        int readerIndex = packetData.readerIndex();
+        byte[] srcMACAddr = new byte[6];
+        byte[] destMACAddr = new byte[6];
+
+        packetData.getBytes(readerIndex + headerOffset + 6, srcMACAddr);
+        packetData.getBytes(readerIndex + headerOffset + 0, destMACAddr);
+
+        formatter.format("  Ethernet: src ");
+        formatMAC48(srcMACAddr, formatter);
+        formatter.format(", dest ");
+        formatMAC48(destMACAddr, formatter);
+        formatter.format(", protocol %s\n", TunTapPacket.getProtocolName(protocol));
+    }
+
+    private static void formatIPv4Header(ByteBuf packetData, int headerOffset, Formatter formatter) {
+        int readerIndex = packetData.readerIndex();
+        String srcAddr, destAddr;
+
+        try {
+            byte[] srcAddrData = new byte[4];
+            packetData.getBytes(readerIndex + headerOffset + 12, srcAddrData);
+            srcAddr = InetAddress.getByAddress(srcAddrData).getHostAddress();
+        } catch (Exception ex) {
+            srcAddr = "(invalid)";
+        }
+
+        try {
+            byte[] destAddrData = new byte[4];
+            packetData.getBytes(readerIndex + headerOffset + 16, destAddrData);
+            destAddr = InetAddress.getByAddress(destAddrData).getHostAddress();
+        } catch (Exception ex) {
+            destAddr = "(invalid)";
+        }
+
+        int transportProtocol = packetData.getUnsignedByte(readerIndex + headerOffset + 9);
+
+        formatter.format("  IPv4: src %s, dest %s, protocol %s\n",
+                srcAddr, destAddr, transportProtocolToString(transportProtocol));
+    }
+
+    private static void formatIPv6Header(ByteBuf packetData, int headerOffset, Formatter formatter) {
+        int readerIndex = packetData.readerIndex();
+        String srcAddr, destAddr;
+
+        try {
+            byte[] srcAddrData = new byte[16];
+            packetData.getBytes(readerIndex + headerOffset + 8, srcAddrData);
+            srcAddr = InetAddress.getByAddress(srcAddrData).getHostAddress();
+        } catch (Exception ex) {
+            srcAddr = "(invalid)";
+        }
+
+        try {
+            byte[] destAddrData = new byte[16];
+            packetData.getBytes(readerIndex + headerOffset + 24, destAddrData);
+            destAddr = InetAddress.getByAddress(destAddrData).getHostAddress();
+        } catch (Exception ex) {
+            destAddr = "(invalid)";
+        }
+
+        int transportProtocol = packetData.getUnsignedByte(readerIndex + headerOffset + 6);
+
+        formatter.format("  IPv6: src %s, dest %s, protocol %s\n",
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) %n should be used in place of \n to produce the platform-specific line separator. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2275)